Add config_flow to AndroidTV integration

ollo69 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Breaking change

Config flow support has been added to `androidtv`. Existing YAML configuration entries will be imported automatically but should be removed from YAML configuration after import. Import support will be removed in a future release.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This PR is to implement Configuration flow for AndroidTV integration.
Entry and entity `unique_id` now are based on device Mac Address. Migration of old entities with `unique_id` based on device SerialNo is implemented. 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

ollo69 committed:
Add Config Flow to AndroidTV

ollo69 committed:
Variable rename

ollo69 committed:
Fix and add AndroidTV tests

ollo69 committed:
Add check for configured host

ollo69 committed:
Implemented migration from yaml

ollo69 committed:
Replace array with tuple

JeffLIrion commented on homeassistant/components/androidtv/config_flow.py:
```diff
@@ -0,0 +1,309 @@
+        return False
+    if not os.access(file_in, os.R_OK):
+        return False
+    return True
```
```python
return os.path.isfile(file_in) and os.access(file_in, os.R_OK):
```

ollo69 replied:
Done

JeffLIrion commented on homeassistant/components/androidtv/config_flow.py:
```diff
@@ -0,0 +1,309 @@
+
+
+class OptionsFlowHandler(config_entries.OptionsFlow):
+    """Handle a option flow for Android TV."""
```
```suggestion
    """Handle an option flow for Android TV."""
```

JeffLIrion commented on homeassistant/components/androidtv/const.py:
```diff
@@ -0,0 +1,31 @@
+ANDROID_DEV = DOMAIN
+ANDROID_DEV_OPT = "androidtv_opt"
+
+CONF_ADB_KEY = "adb_key"
```
I used "adbkey" because the key files are named `adbkey` and `adbkey.pub`. I would prefer to stick to that, and also make this constant `CONF_ADBKEY`.

ollo69 replied:
Reverted name as requested. BTW I was also breaking the YAML configuration.😊

JeffLIrion commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -449,6 +345,68 @@ def __init__(
+        sw_version = info.get("sw_version")
+        mac = info.get("ethmac")
+        if not mac:
+            mac = info.get("wifimac")
```
```python
            mac = info.get("ethmac") or info.get("wifimac")
```

tkdrob commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -168,180 +150,104 @@
-    return adbkey, signer, adb_log
-
-
 async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
```
```suggestion
async def async_setup_platform(hass: HomeAssistant, config: ConfigType, async_add_entities, discovery_info=None):
```

ollo69 replied:
Done, also for `async_add_entities`

tkdrob commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -33,25 +30,49 @@
-ANDROIDTV_DOMAIN = "androidtv"
+from homeassistant.helpers.device_registry import CONNECTION_NETWORK_MAC
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+
```
```suggestion
from homeassistant.helpers.typing import ConfigType

```

tkdrob commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -85,53 +106,12 @@
-
+DEFAULT_NAME = "Android TV"
 
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
```
```suggestion
# Deprecated in Home Assistant 2021.9
PLATFORM_SCHEMA = cv.deprecated(
    vol.All(
        PLATFORM_SCHEMA.extend(
```

ollo69 replied:
Done

tkdrob commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -168,180 +150,104 @@
-    # Close the ADB connection when HA stops
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _async_close)
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities
```
```suggestion
    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: entity_platform.AddEntitiesCallback
```

ollo69 replied:
Done

tkdrob commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -153,7 +133,9 @@
+            CONF_EXCLUDE_UNNAMED_APPS, default=DEFAULT_EXCLUDE_UNNAMED_APPS
+        ): cv.boolean,
         vol.Optional(CONF_SCREENCAP, default=DEFAULT_SCREENCAP): cv.boolean,
     }
 )
```
```suggestion
            }
        ),
    )
)
```

tkdrob commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -168,180 +150,104 @@
+    config_data = {
+        CONF_HOST: host,
+        CONF_NAME: config[CONF_NAME],
+        CONF_DEVICE_CLASS: config[CONF_DEVICE_CLASS],
+        CONF_PORT: config[CONF_PORT],
+        CONF_ADB_SERVER_PORT: config[CONF_ADB_SERVER_PORT],
```
```suggestion
        CONF_DEVICE_CLASS: config.get(CONF_DEVICE_CLASS),
        CONF_PORT: config.get(CONF_PORT),
        CONF_ADB_SERVER_PORT: config.get(CONF_ADB_SERVER_PORT),
```

ollo69 replied:
Done, also fixed the part related to the options.

tkdrob reviewed (commented):
We need to be sure that import occurs with a minimal config and key errors be handled.

ollo69 committed:
Requested changes

ollo69 committed:
Const isort

JeffLIrion commented on homeassistant/components/androidtv/const.py:
```diff
@@ -0,0 +1,31 @@
+"""Android TV component constants."""
+DOMAIN = "androidtv"
+
+ANDROID_DEV = DOMAIN
```
Why do you have two constants that are the same? Do they have different meanings, even though their values are the same?

ollo69 replied:
Yes, I use `ANDROID_DEV` to store the device created in `hass` data and retrieve it during media player entity initialization.
This is what I do:

```
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
        ANDROID_DEV: aftv,
        ANDROID_DEV_OPT: entry.options.copy(),
        "conf_listener": entry.add_update_listener(update_listener),
        "stop_listener": stop_listener,
    }
```

So they mean 2 different things, but the value can be the same.

ollo69 commented:
> The integration currently uses the serial number as the unique ID.

I'm already changing unique ID for the entity, because I'm adding `media-player` suffix, and this could be a problem because the entity name will be probably renamed after migration and this should be considered a breaking change. I could remove this suffix for the entity, but if the serial number cannot be used as unique ID for the entry the problem cannot be solved.
So at the end the best option could be to concatenate serial and mac-address and mark this as breaking change. At this point it will be probably better to use a different naming convention during migration to avoid creation of entityID with _1 suffix? E.g. concatenate `device type` and `host` and ignore the `name` parameter? Let's wait for other opinion here.

In any case the most important question is if the library property `device_properties` is reliable. It will not be possible to create the entry if is not possible to determinate a valid unique ID, whatever we use for it

tkdrob commented:
> > I used the `SerialNo` of the device get from `device_properties` to define entity `unique_id`, but I don't know if this methods always works (it works with my fireTV).
> 
> The integration currently uses the serial number as the unique ID. However, there is an unresolved bug where some devices reuse the same serial number (which is really a bug on the manufacturer's end), and that prevents setting up a second device. @balloob suggested using the MAC address as the unique ID ([link](https://github.com/home-assistant/core/issues/35325#issuecomment-625920341)). The only potential problem that I see with that approach is that it would require migrating existing instances to the new unique ID. But that might be unavoidable because changing what's used for the unique ID seems to be the only way to fix that issue.
> 
> I don't know if there would be any benefit in concatenating the serial number and MAC address in order to form the unique ID.

If the mac address is available from the api, I believe that is all you need, no concatenating. That way DHCP being added later will not trigger flows.

ollo69 committed:
Use mac as unique ID and remove CONF_NAME

ollo69 commented:
> If the mac address is available from the api, I believe that is all you need, no concatenating. That way DHCP being added later will not trigger flows.

I changed unique ID to use mac address. I also removed the import of the `name` configured in the yaml configuration (user can rename the entity from UI) and added information about the breaking change in the PR.

tkdrob commented on homeassistant/components/androidtv/config_flow.py:
```diff
@@ -0,0 +1,304 @@
+        if errors:
+            return self._show_setup_form(user_input, errors)
+
+        self._async_abort_entries_match({CONF_HOST: self._host})
+        if ip_address != self._host:
+            self._async_abort_entries_match({CONF_HOST: ip_address})
```
Is there a reason why we have to evaluate an address other than what the user specifies? I'm assuming that `ip_address` is what the adb device reports?

tkdrob replied:
```suggestion
            self._async_abort_entries_match({CONF_HOST: host})
            if ip_address != host:
                self._async_abort_entries_match({CONF_HOST: ip_address})
```

JeffLIrion commented on homeassistant/components/androidtv/__init__.py:
```diff
@@ -1 +1,195 @@
+
+
+async def async_connect_androidtv(
+    hass, config, *, state_detection_rules=None, timeout=10.0
```
10.0 means that the user has 10 seconds to accept the connection on their Android device. It's fine if they don't accept it within that window, the next time it tries to connect the key will be pre-approved. But I think the config flow would fail. Should a larger timeout be used?

ollo69 replied:
This is already implemented. In the config flow the call to method `async_connect_androidtv` is done using `timeout=30.0`

JeffLIrion replied:
In that case, maybe the default value should be 30 instead of 10? When the key pair has already been approved, that parameter isn't used.

ollo69 replied:
In my mind it is better to set a lower default timeout and increase it when required. By example, if the `adbkey` is cancelled, we don't want to wait 30 sec if we are not in config flow. What do you think?

JeffLIrion replied:
I'm not sure what you mean by "if the `adbkey` is cancelled." Do you mean that the user clears the trusted keys on their Android device? Or if the user rejects the connection when the pop-up appears on their TV? The latter would be more likely to happen during config flow, but could happen on any reconnect attempt if the user didn't select "always trust connections from this computer."

Either way, I think it's an unlikely scenario. And if the user does something like that, I think it falls outside of the intended usage and having to wait 30 seconds instead of 10 is not unreasonable. 

I would view that timeout as

* If you're connecting for the first time (in config flow), you have 30 seconds to approve the connection or the config flow will fail and you'll need to repeat it
* In all other scenarios, the ADB key should be pre-approved and that timeout has no effect, so it doesn't matter if it's 10 or 30

It's not a big deal, though.

ollo69 replied:
May be I'm wrong, but I was thinking about a situation where the file with `adbkey` is physically deleted from HA folder and HA automatically generate a new one at restart. At this point the key is changed and so the device will pop-up again  for authorization, doesn't it?
Anyway timeout is already as described by you, 30 seconds for config flow, 10 seconds in all other scenario, that should not matter because key should be pre-approved.
But as you said, this is not a big deal, if you think that is better to have 30 sec, in all scenario I will change it for all, I leave final decision on your hand.😉

JeffLIrion replied:
I would use 30 seconds for all scenarios.

ollo69 replied:
Set timeout 30 seconds for all scenarios (default value)

tkdrob commented on homeassistant/components/androidtv/config_flow.py:
```diff
@@ -0,0 +1,304 @@
+        if result != RESULT_SUCCESS:
+            return self._show_setup_form(user_input, {"base": result})
+
+        if not self._unique_id:
+            return self.async_abort(reason="invalid_unique_id")
```
In what cases would the tv not return a mac address?

JeffLIrion replied:
Here are the commands used to get the MAC addresses: https://github.com/JeffLIrion/python-androidtv/blob/ebbcfafb194fb0e022d37f5ecb55d8d9dc8d0735/androidtv/constants.py#L124-L126

Here is some example output: https://github.com/JeffLIrion/python-androidtv/blob/ebbcfafb194fb0e022d37f5ecb55d8d9dc8d0735/tests/test_basetv_sync.py#L18-L71

JeffLIrion replied:
^^^You can test those commands on your device(s) to make sure that there is at least one MAC address. 

But maybe it's worth concatenating the MAC address and the serial number, just in case neither of those ADB commands returns a MAC address?

tkdrob replied:
```suggestion
                if not unique_id:
                    return self.async_abort(reason="invalid_unique_id")
```

JeffLIrion commented:
> ## Breaking change
> Because after import the `entity name` will change, it will be necessary to re-configure any `automation` that use old `entity name`.

You could also add that the user can change the entity ID in the UI back to its previous value. That's probably the easiest fix.

> ## Proposed change
> I used the `SerialNo` of the device get from `device_properties` to define entity `unique_id`, but I don't know if this methods always works (it works with my fireTV).

This is no longer accurate, although I don't know that it's 100% decided that the MAC address will be used.

ollo69 commented:
> This is no longer accurate, although I don't know that it's 100% decided that the MAC address will be used.

Agree, but i will wait for final decision before edit this. BTW an admin should add the `Breaking change` labels.

tkdrob commented on homeassistant/components/androidtv/__init__.py:
```diff
@@ -1 +1,195 @@
+        config[CONF_HOST],
+        config[CONF_PORT],
+        adbkey,
+        config.get(CONF_ADB_SERVER_IP, ""),
+        config[CONF_ADB_SERVER_PORT],
+        state_detection_rules,
+        config[CONF_DEVICE_CLASS],
```
```suggestion
        config.get(CONF_ADB_SERVER_IP),
        config.get(CONF_ADB_SERVER_PORT),
        state_detection_rules,
        config.get(CONF_DEVICE_CLASS),
```

ollo69 replied:
I don't think that we can avoid breaking change. The problem is that the entity is currently using an unique ID based on serial No, because we are going to change unique ID (based on mac) a new entity with new entity ID will be created.

JeffLIrion replied:
I know very little about config flow stuff, but @balloob previously said ([link](https://github.com/home-assistant/core/issues/35325#issuecomment-625920341)):

> It doesn't have to be a breaking change because we can do a migration. `async_migrate_entries` exists in `helpers/entity_registry.py`.

I haven't looked at @tkdrob's suggested changes, so for all I know this might be what he's doing.

ollo69 replied:
I don't see any migration here, this seems more related to simplify config flow. I start merging this changes, we can come back to breaking change later.

tkdrob commented on homeassistant/components/androidtv/config_flow.py:
```diff
@@ -0,0 +1,304 @@
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_DEVICE_CLASS, CONF_HOST, CONF_PORT
+from homeassistant.core import callback
+from homeassistant.helpers import config_validation as cv
```
```suggestion
from homeassistant.const import CONF_DEVICE_CLASS, CONF_HOST, CONF_NAME, CONF_PORT
from homeassistant.core import callback
from homeassistant.helpers import config_validation as cv
from homeassistant.helpers.device_registry import format_mac
```

tkdrob commented on homeassistant/components/androidtv/config_flow.py:
```diff
@@ -0,0 +1,304 @@
+
+    VERSION = 1
+
+    def __init__(self):
+        """Initialize Android TV config flow."""
+        self._host = None
+        self._unique_id = None
```
```suggestion
```

tkdrob commented on homeassistant/components/androidtv/config_flow.py:
```diff
@@ -0,0 +1,304 @@
+        self._unique_id = None
+
+    @callback
+    def _show_setup_form(self, user_input=None, errors=None):
+        """Show the setup form to the user."""
+
+        if user_input is None:
+            user_input = {}
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_HOST, default=user_input.get(CONF_HOST, "")): str,
+                    vol.Required(
+                        CONF_DEVICE_CLASS, default=DEFAULT_DEVICE_CLASS
+                    ): vol.In(DEVICE_CLASSES),
+                    vol.Required(CONF_PORT, default=DEFAULT_PORT): cv.port,
+                    vol.Optional(CONF_ADBKEY): str,
+                    vol.Optional(CONF_ADB_SERVER_IP): str,
+                    vol.Required(
+                        CONF_ADB_SERVER_PORT, default=DEFAULT_ADB_SERVER_PORT
+                    ): cv.port,
+                }
+            ),
+            errors=errors or {},
+        )
```
```suggestion
    def _show_setup_form(self, user_input=None, error=None):
        """Show the setup form to the user."""
        user_input = user_input or {}
        data_schema = {
            vol.Required(CONF_HOST, default=user_input.get(CONF_HOST, "")): str,
            vol.Required(CONF_DEVICE_CLASS, default=DEFAULT_DEVICE_CLASS): vol.In(
                DEVICE_CLASSES
            ),
            vol.Required(CONF_PORT, default=DEFAULT_PORT): cv.port,
        }

        if self.show_advanced_options:
            data_schema = {
                vol.Required(CONF_HOST, default=user_input.get(CONF_HOST, "")): str,
                vol.Required(CONF_DEVICE_CLASS, default=DEFAULT_DEVICE_CLASS): vol.In(
                    DEVICE_CLASSES
                ),
                vol.Required(CONF_PORT, default=DEFAULT_PORT): cv.port,
                vol.Optional(CONF_ADBKEY): str,
                vol.Optional(CONF_ADB_SERVER_IP): str,
                vol.Required(
                    CONF_ADB_SERVER_PORT, default=DEFAULT_ADB_SERVER_PORT
                ): cv.port,
            }

        return self.async_show_form(
            step_id="user",
            data_schema=vol.Schema(data_schema),
            errors={"base": error},
        )
```

tkdrob replied:
This could probably be simplified a bit. Since the adb server addon is deprecated, only advanced users should see this. Perhaps some warning should be displayed to the user to suggest moving away from adb server.

tkdrob commented on homeassistant/components/androidtv/config_flow.py:
```diff
@@ -0,0 +1,304 @@
+            aftv = await async_connect_androidtv(self.hass, user_input, timeout=30.0)
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception(
+                "Unknown error connecting with Android TV at %s", self._host
+            )
+            return RESULT_UNKNOWN
+
+        if not aftv:
+            return RESULT_CONN_ERROR
+
+        dev_prop = aftv.device_properties
+        self._unique_id = dev_prop.get("ethmac") or dev_prop.get("wifimac")
+        await aftv.adb_close()
+        return RESULT_SUCCESS
```
```suggestion
                "Unknown error connecting with Android TV at %s", user_input[CONF_HOST]
            )
            return RESULT_UNKNOWN, None

        if not aftv:
            return RESULT_CONN_ERROR, None

        dev_prop = aftv.device_properties
        unique_id = format_mac(dev_prop.get("ethmac", "wifimac"))
        await aftv.adb_close()
        return None, unique_id
```

ollo69 replied:
```
       unique_id = format_mac(dev_prop.get("ethmac", "wifimac"))
```

This is wrong, we want to get  the item with ID "wifimac" if the one with "ethmac" is not available, not set "wifimac" as default

tkdrob commented on homeassistant/components/androidtv/config_flow.py:
```diff
@@ -0,0 +1,304 @@
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initiated by the user."""
+        if user_input is None:
+            return self._show_setup_form(user_input)
+
+        errors = {}
+        self._unique_id = None
+        self._host = user_input[CONF_HOST]
+        adb_key = user_input.get(CONF_ADBKEY)
+        adb_server = user_input.get(CONF_ADB_SERVER_IP)
```
```suggestion
         error = None
        if user_input is not None:
            host = user_input[CONF_HOST]
            adb_key = user_input.get(CONF_ADBKEY)
            adb_server = user_input.get(CONF_ADB_SERVER_IP)
```

tkdrob commented on homeassistant/components/androidtv/config_flow.py:
```diff
@@ -0,0 +1,304 @@
+        adb_key = user_input.get(CONF_ADBKEY)
+        adb_server = user_input.get(CONF_ADB_SERVER_IP)
+
+        if adb_key and adb_server:
+            errors["base"] = "key_and_server"
+        elif adb_key:
+            isfile = await self.hass.async_add_executor_job(_is_file, adb_key)
+            if not isfile:
+                errors["base"] = "adbkey_not_file"
+
+        if not errors:
+            ip_address = await self.hass.async_add_executor_job(_get_ip, self._host)
+            if not ip_address:
+                errors["base"] = "invalid_host"
+
+        if errors:
+            return self._show_setup_form(user_input, errors)
```
```suggestion
            if adb_key and adb_server:
                return self._show_setup_form(user_input, "key_and_server")
            if adb_key:
                isfile = await self.hass.async_add_executor_job(_is_file, adb_key)
                if not isfile:
                    return self._show_setup_form(user_input, "adbkey_not_file")

            ip_address = await self.hass.async_add_executor_job(_get_ip, host)
            if not ip_address:
                return self._show_setup_form(user_input, "invalid_host")
```

tkdrob commented on homeassistant/components/androidtv/config_flow.py:
```diff
@@ -0,0 +1,304 @@
+        if ip_address != self._host:
+            self._async_abort_entries_match({CONF_HOST: ip_address})
+
+        result = await self._async_check_connection(user_input)
+        if result != RESULT_SUCCESS:
+            return self._show_setup_form(user_input, {"base": result})
+
```
```suggestion
            error, unique_id = await self._async_check_connection(user_input)
            if error is None:
```

tkdrob commented on homeassistant/components/androidtv/config_flow.py:
```diff
@@ -0,0 +1,304 @@
+        if not self._unique_id:
+            return self.async_abort(reason="invalid_unique_id")
+
+        await self.async_set_unique_id(self._unique_id)
+        self._abort_if_unique_id_configured()
+
+        return self.async_create_entry(
+            title=self._host,
+            data=user_input,
+        )
```
```suggestion
                await self.async_set_unique_id(unique_id)
                self._abort_if_unique_id_configured()

                return self.async_create_entry(
                    title=user_input.get(CONF_NAME) or host,
                    data=user_input,
                )
        user_input = user_input or {}
        return self._show_setup_form(user_input, error)
```

tkdrob commented on homeassistant/components/androidtv/config_flow.py:
```diff
@@ -0,0 +1,304 @@
+            data=user_input,
+        )
+
+    async def async_step_import(self, user_input=None):
+        """Import a config entry."""
+        return await self.async_step_user(user_input)
```
```suggestion
    async def async_step_import(self, import_config=None):
        """Import a config entry."""
        return await self.async_step_user(import_config)
```

tkdrob commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -168,180 +159,108 @@
-            "Could not connect to %s at %s %s", device_name, address, adb_log
+    # get main data
+    config_data = {
+        CONF_HOST: host,
```
```suggestion
        CONF_HOST: host,
        CONF_NAME: config.get(CONF_NAME, DEFAULT_NAME),
```

tkdrob commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -168,180 +159,108 @@
+    device_class = aftv.DEVICE_CLASS
+    device_name = "Android TV " if device_class == DEVICE_ANDROIDTV else "Fire TV "
+    device_name += entry.data[CONF_HOST]
 
```
```suggestion
    if CONF_NAME in entry.data:
        device_name = entry.data[CONF_NAME]

```

ollo69 replied:
Because new entity will be created due to different Unique ID, make sense try to assign the same name? I think that this will only cause a duplicated entity with "_1" suffix

tkdrob commented on homeassistant/components/androidtv/strings.json:
```diff
@@ -0,0 +1,57 @@
+        "description": "Set required parameter to connect to your Android TV device",
+        "data": {
+          "host": "[%key:common::config_flow::data::host%]",
+          "adb_key": "Path to your ADB key file (leave empy to auto generate)",
```
```suggestion
          "adbkey": "Path to your ADB key file, leave empty to auto generate",
```

tkdrob commented on homeassistant/components/androidtv/strings.json:
```diff
@@ -0,0 +1,57 @@
+    },
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]",
+      "invalid_unique_id": "Impossible to determinate a valid unique id for the device"
```
```suggestion
      "invalid_unique_id": "Impossible to determine a valid unique id for the device"
```

tkdrob commented on homeassistant/components/androidtv/translations/en.json:
```diff
@@ -0,0 +1,57 @@
+    "config": {
+        "abort": {
+            "already_configured": "Device is already configured",
+            "invalid_unique_id": "Impossible to determinate a valid unique id for the device"
```
```suggestion
            "invalid_unique_id": "Impossible to determine a valid unique id for the device"
```

tkdrob commented on homeassistant/components/androidtv/translations/en.json:
```diff
@@ -0,0 +1,57 @@
+        "step": {
+            "user": {
+                "data": {
+                    "adb_key": "Path to your ADB key file (leave empy to auto generate)",
```
```suggestion
                    "adbkey": "Path to your ADB key file, leave empty to auto generate",
```

tkdrob commented on tests/components/androidtv/test_config_flow.py:
```diff
@@ -0,0 +1,429 @@
+    """Test when we have errors connecting the router."""
+    flow_result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={"source": SOURCE_USER},
```
```suggestion
        context={"source": SOURCE_USER, "show_advanced_options": True},
```

tkdrob reviewed (changes requested):
These suggested changes should allow this to not be a breaking change. @bieniu A side not, it seems that the original code did not always distinguish between android tvs and fire tvs so the default name was `android_tv` during my testing.

ollo69 committed:
Apply suggestions from code review

Co-authored-by: Robert Hillis <tkdrob4390@yahoo.com>

ollo69 committed:
Set default timeout to 30 sec and minor change

tkdrob reviewed (commented):
This could work for "migration" but it's a bit dirty. This is still breaking as the user could have changed the entity id in the UI which wins over everything else. I haven't tested multiple devices together. We will also need the name users have specified in yaml which is what was suggested already.

Edit: This seems to actually work quite well, now. We update the existing entity in the registry which keeps everything the user has done with it up to this point.

ollo69 committed:
Implement migration for old entities

ollo69 commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -168,180 +159,107 @@
+        entry.entry_id,
+        entry.unique_id,
     ]
 
```
I implemented a different approach, trying to migrate existing entity with old unique ID, but may be yours solution is better. Please check and let me know.

tkdrob replied:
Your approach works the same as mine. I think the members will want your approach if it uses `async def async_migrate_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:` That way HA knows it supports migration.

ollo69 replied:
I think that I cannot use `async_migrate_entry` here because the `entity` is still not related to the `entry`. 
Anyway this seems working a could avoid breaking change, let's way for members review.

tkdrob replied:
My mistake. Take a look here: https://github.com/home-assistant/core/blob/dev/homeassistant/components/powerwall/__init__.py#L47

tkdrob replied:
My mistake. Take a look here: https://github.com/home-assistant/core/blob/dev/homeassistant/components/powerwall/__init__.py#L47

ollo69 replied:
Yes, I already see this code. But `async_migrate_entry` try to upgrade entity that are already related to the entry. Because the relation is created on first setup, it does not work.

ollo69 committed:
Improve test coverage

tkdrob commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -658,6 +643,11 @@ class FireTVDevice(ADBDevice):
 
     _attr_supported_features = SUPPORT_FIRETV
 
+    @property
+    def device_info(self):
+        """Return a device description for device registry."""
+        return self._get_device_info("FireTV")
+
     @adb_decorator(override_available=True)
     async def async_update(self):
         """Update the device state and, if necessary, re-connect."""
```
```suggestion
    @adb_decorator(override_available=True)
    async def async_update(self):
        """Update the device state and, if necessary, re-connect."""
        self._attr_device_info = self._get_device_info("FireTV")

```

tkdrob commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -586,6 +566,11 @@ class AndroidTVDevice(ADBDevice):
 
     _attr_supported_features = SUPPORT_ANDROIDTV
 
+    @property
+    def device_info(self):
+        """Return a device description for device registry."""
+        return self._get_device_info("AndroidTV")
+
     @adb_decorator(override_available=True)
     async def async_update(self):
         """Update the device state and, if necessary, re-connect."""
```
```suggestion
    @adb_decorator(override_available=True)
    async def async_update(self):
        """Update the device state and, if necessary, re-connect."""
        self._attr_device_info = self._get_device_info("AndroidTV")

```

ollo69 replied:
I think that `device_info` should be called before than first update, in this way was skipping some rows in tests. I restored the the property.

tkdrob replied:
What if you add `,True` to `async_add_entities`? Does it work then?

ollo69 replied:
Think is better to not change this, but I found a different solution to use attribute.

ollo69 replied:
Seems that this attribute must be provided before added to hass or device is nor created. Now should be ok,

ollo69 committed:
Apply suggestions from code review

Co-authored-by: Robert Hillis <tkdrob4390@yahoo.com>

ollo69 committed:
Restore device_info prop and fix flake8

ollo69 committed:
Use back _attr_device_info

ollo69 committed:
Fix device property

ollo69 committed:
Remove unuseful attribute

tkdrob commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -1,10 +1,9 @@
 """Support for functionality to interact with Android TV / Fire TV devices."""
 from datetime import datetime
```
```suggestion
from __future__ import annotations

from datetime import datetime
```

tkdrob commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -397,34 +325,24 @@ def __init__(
-        self.turn_on_command = turn_on_command
-        self.turn_off_command = turn_off_command
+        self._attr_unique_id = unique_id
+        self._attr_device_info = self._get_device_info(dev_type, unique_id)
```
```suggestion
        info = aftv.device_properties
        model = info.get(ATTR_MODEL)
        self._attr_device_info = {
            "connections": {
                (
                    CONNECTION_NETWORK_MAC,
                    format_mac(info.get(PROP_ETHMAC) or info.get(PROP_WIFIMAC, "")),
                )
            },
            ATTR_IDENTIFIERS: {(DOMAIN, unique_id)},
            ATTR_MANUFACTURER: info.get(ATTR_MANUFACTURER),
            ATTR_MODEL: f"{model} ({dev_type})" if model else dev_type,
            ATTR_NAME: name,
            ATTR_SW_VERSION: info.get(ATTR_SW_VERSION),
        }
```

ollo69 replied:
It is correct set the `dict` items also if they are empty (e.g. `ATTR_SW_VERSION: info.get(ATTR_SW_VERSION)`)?

tkdrob replied:
If version is available, it will display in the device info. If not, version will not be displayed at all.

ollo69 replied:
Also refer to `model` and `connections` (we could have `"connections": {CONNECTION_NETWORK_MAC, ""`)

tkdrob commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -449,6 +367,66 @@ def __init__(
+            )
+        )
+        return
+
+    def _get_device_info(self, dev_type, dev_id):
+        """Get device information."""
+        info = self.aftv.device_properties
+        model = info.get(ATTR_MODEL)
+        model = f"{model} ({dev_type})" if model else dev_type
+        manufacturer = info.get(ATTR_MANUFACTURER)
+        sw_version = info.get(ATTR_SW_VERSION)
+        mac = format_mac(info.get(PROP_ETHMAC) or info.get(PROP_WIFIMAC, ""))
+
+        data = {
+            ATTR_IDENTIFIERS: {(DOMAIN, dev_id)},
+            ATTR_NAME: self.name,
+            ATTR_MODEL: model,
+        }
+        if manufacturer:
+            data[ATTR_MANUFACTURER] = manufacturer
+        if mac:
+            data["connections"] = {(CONNECTION_NETWORK_MAC, mac)}
+        if sw_version:
+            data[ATTR_SW_VERSION] = sw_version
+        return data
```
```suggestion
```

tkdrob commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -449,6 +367,66 @@ def __init__(
+        return data
+
     @property
     def media_image_hash(self):
```
```suggestion
    def media_image_hash(self) -> str | None:
```

tkdrob commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -397,34 +325,24 @@ def __init__(
+        self.turn_off_command = None
 
         # ADB exceptions to catch
         if not self.aftv.adb_server_ip:
```
```suggestion
        if not aftv.adb_server_ip:
```

ollo69 committed:
Apply suggestions from code review

Co-authored-by: Robert Hillis <tkdrob4390@yahoo.com>

ollo69 committed:
Check mac availaible for device propertiies

ollo69 committed:
Minor change for device info

JeffLIrion commented on homeassistant/components/androidtv/__init__.py:
```diff
@@ -1 +1,232 @@
+
+    if CONF_ADB_SERVER_IP in config:
+        _LOGGER.warning(
+            "Adb server is deprecated. Please consider connecting directly to your TV"
```
The ADB add-on is deprecated, but using an ADB server in the `androidtv` library is not -- I have no plans to remove that functionality. 

If a user wants to maintain their own ADB server, that's perfectly fine. But if the Python ADB implementation (`adb-shell`) works for that user then I don't know why they would go to that trouble. An ADB server also makes debugging much more challenging. 

I'll leave it up to you to decide if you want to keep, reword, or remove this log statement.

ollo69 replied:
In this case I think that log should be removed or at least moved to `info` level.

JeffLIrion replied:
I suggest rewording it to state that the ADB add-on is deprecated, not the ADB server config option.

If you really want to get fancy, I believe there's a way to check if an add-on is installed, so you could only log this if the ADB add-on is installed. But I'm sure the vast majority of people using an ADB server are utilizing the add-on, so it might be more trouble than it's worth to do that.

tkdrob replied:
Perhaps we should remove it then. I can't easily get the addon installed for testing and I'm note sure what the exact addon slug is called for `async_get_addon_info`.

ollo69 replied:
I removed this warning. I suppose that is the obsolete plugin that should raise it.

tkdrob commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -85,77 +110,45 @@
+    )
 )
 
 # Translate from `AndroidTV` / `FireTV` reported state to HA state.
```
I would have `ANDROIDTV_STATES` use state constants rather than strings.

JeffLIrion replied:
The keys in this dictionary are strings in the `androidtv` package: https://github.com/JeffLIrion/python-androidtv/blob/7d47c3a35f247cd03bdc9ad0d9653e428edeb204/androidtv/constants.py#L275-L282

ollo69 replied:
I think we should avoid to change things that have nothing to do with Config Flow in this PR. This could be a new PR.

tkdrob reviewed (approved):
Looks good. Hopefully a member can review this soon.

ollo69 committed:
Remove warning for adb_server usage

ollo69 committed:
Translations fix

ollo69 committed:
Replace "defaut" with "suggested_value" in options

ollo69 committed:
Minor code clean-up

ollo69 committed:
Resolve conflict and use DeviceInfo

ollo69 committed:
Fix on entity migrate method

tkdrob commented:
Any idea why Android 11 does not work properly? That may be the reason why this PR hasn't got member attention yet.

ollo69 commented:
I wasn't aware about this, I just have 2 fire TV stick that works without issue. But I suppose that is an API issue, nothing to do with integration and related config flow. I just suppose that member team is too busy to follow a big PR like this.

jherby2k commented:
I don't see what the android 11 issue has to do with this PR. Its a separate issue that will need to be addressed in a different PR. This PR improves the quality of this component and will allow for new options currently blocked by the yaml ban.

bdraco commented on homeassistant/components/androidtv/config_flow.py:
```diff
@@ -0,0 +1,302 @@
+            }
+        )
+        if app_id == APPS_NEW_ID:
+            data_schema = data_schema.extend(
+                {
+                    vol.Optional(CONF_APP_ID): str,
+                }
+            )
```
```suggestion
            data_schema[vol.Optional(CONF_APP_ID)] = str 
```

bdraco replied:
If you make everything a `dict`, and then wrap it in `vol.Schema` at the end, this will work.

bdraco commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -450,8 +384,46 @@ def __init__(
+    def _process_config(self):
+        """Load the config options."""
+        _LOGGER.debug("Loading configuration options")
+        options = self.hass.data[DOMAIN][self._entry_id][ANDROID_DEV_OPT]
```
It would be better to pass in the options to the constructor instead of accessing `hass.data` in the entity if it can be avoided

bdraco replied:
It looks like you probably can't do that since you change it here : https://github.com/home-assistant/core/pull/54444/files#diff-99e4bd484168f8f14de01521b3083aa44bdf3484b9283d2a7cd4f8bb1210eefeR227

bdraco replied:
You could pass `self.hass.data[DOMAIN][self._entry_id]` into the constructor instead since it will still have a reference to it

bdraco commented on homeassistant/components/androidtv/config_flow.py:
```diff
@@ -0,0 +1,302 @@
+            adb_server = user_input.get(CONF_ADB_SERVER_IP)
+
+            if adb_key and adb_server:
+                return self._show_setup_form(user_input, "key_and_server")
```
Move to a new step in this case

bdraco commented on homeassistant/components/androidtv/config_flow.py:
```diff
@@ -0,0 +1,302 @@
+
+        if user_input is not None:
+            host = user_input[CONF_HOST]
+            adb_key = user_input.get(CONF_ADBKEY)
```
Since we have the host we should validate it before moving on

bdraco commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -168,180 +168,111 @@
+            else:
+                config_options[key] = config[key]
+
+    # save option to use with entry
+    if config_options:
+        hass.data.setdefault(DOMAIN, {}).setdefault(MIGRATION_DATA, {})[
+            host
+        ] = config_options
```
Instead of storing this in `hass.data` pass it on to the config flow

bdraco commented on tests/components/androidtv/patchers.py:
```diff
@@ -153,6 +153,7 @@ def isfile(filepath):
 
 PATCH_ISFILE = patch("os.path.isfile", isfile)
 PATCH_ACCESS = patch("os.access", return_value=True)
+PATCH_GET_HOST_IP = patch("socket.gethostbyname", return_value="127.0.0.1")
```
Ideally we patch only the location of one we want to mock and not the global one

ollo69 replied:
I do not exactly understand what you mean here.

tkdrob replied:
He wants to see socket patched as it is imported

bdraco commented on tests/components/androidtv/test_config_flow.py:
```diff
@@ -0,0 +1,446 @@
+        context={"source": SOURCE_USER, "show_advanced_options": True},
+    )
+
+    with patch(
+        CONNECT_METHOD,
+        return_value=None,
+    ), PATCH_GET_HOST_IP:
+        result = await hass.config_entries.flow.async_configure(
+            flow_result["flow_id"], user_input=CONFIG_ADB_SERVER
+        )
+        assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+        assert result["errors"] == {"base": "cannot_connect"}
+
+    with patch(
+        CONNECT_METHOD,
+        side_effect=TypeError,
+    ), PATCH_GET_HOST_IP:
+        result = await hass.config_entries.flow.async_configure(
+            flow_result["flow_id"], user_input=CONFIG_ADB_SERVER
+        )
+        assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+        assert result["errors"] == {"base": "unknown"}
```
Follow the config flow all the way through to make sure it can recover from the error and still setup

bdraco reviewed (dismissed):
Looks good.  Just some small suggestions

Please see comments above.

ollo69 committed:
Requested changes

ollo69 committed:
Fix test patch location

bdraco commented on homeassistant/components/androidtv/__init__.py:
```diff
@@ -1 +1,221 @@
+    except ValueError as exp:
+        _LOGGER.warning("Migration of old entity failed: %s", exp)
+
+
+async def async_setup(hass, config):
+    """Set up the Android TV integration."""
+    return True
+
```
```suggestion
```

We can drop this

ollo69 replied:
As I know we cannot drop this because it is required to run `async_setup_platform` that is still there for the import flow.

bdraco replied:
It shouldn't be needed if you have `async_setup_entry` defined after https://github.com/home-assistant/core/pull/48381

ollo69 replied:
If I have a yaml configuration and I run Home Assistant without this block I receive this error:

```
[homeassistant.setup] Setup failed for androidtv: Integration 'androidtv' did not return boolean if setup was successful. Disabling component.
```

and obviously import fail.

tkdrob replied:
When was the last time you merged dev into your branch for this PR? Maybe that has something to do with it?

ollo69 replied:
It is merged with last available dev release and I rebased this PR just yesterday

bdraco commented on homeassistant/components/androidtv/__init__.py:
```diff
@@ -1 +1,221 @@
+        """Close Android TV connection on HA Stop."""
+        await aftv.adb_close()
+
+    stop_listener = hass.bus.async_listen_once(
+        EVENT_HOMEASSISTANT_STOP, async_close_connection
+    )
```
You can wrap this in `entry.async_on_unload` instead.  Same with the `add_update_listener`.  Then you don't have to store it in `hass.data`

bdraco commented on homeassistant/components/androidtv/__init__.py:
```diff
@@ -1 +1,221 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+    if unload_ok:
```
```suggestion
    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
```

bdraco commented on homeassistant/components/androidtv/__init__.py:
```diff
@@ -1 +1,221 @@
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
+        ANDROID_DEV: aftv,
+        ANDROID_DEV_OPT: entry.options.copy(),
+        "conf_listener": entry.add_update_listener(update_listener),
+        "stop_listener": stop_listener,
```
```suggestion
```

Drop this once wrapped in `entry.async_on_unload`

bdraco commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -168,180 +168,110 @@
+    ):
+        if key in config:
+            if key == CONF_STATE_DETECTION_RULES:
+                config_options[key] = json.dumps(config[key])
```
Why is this stored as a json string?

ollo69 replied:
This is a quite complex configuration (see [here](https://www.home-assistant.io/integrations/androidtv/#custom-state-detection)). The only way that I found to manage this as config option is to save it as json string and validating the content with a function provided by the library.

bdraco replied:
One alternative would be to let them provide a state file, but at that point its seems better to leave them in yaml.

Let's ask for another set of eyes on this.

ollo69 committed:
Requested changes

tkdrob commented on homeassistant/components/androidtv/__init__.py:
```diff
@@ -1 +1,216 @@
+        hass, entry.data, state_detection_rules=json_rules
+    )
+    if not aftv:
+        raise ConfigEntryNotReady()
```
Shouldn't a string describing this raised error be included to tell the user why the config entry is not ready?

ollo69 replied:
There is a already a warning message generated by method `async_connect_androidtv`

bdraco replied:
Thats a good point.  The string here will show on the integrations page. The user will have to dig for the log entry.

ollo69 replied:
Ok, changed

ollo69 committed:
Move error message in ConfigEntryNotReady

tkdrob commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -85,77 +117,45 @@
-        vol.Optional(CONF_TURN_OFF_COMMAND): cv.string,
-        vol.Optional(CONF_STATE_DETECTION_RULES, default={}): vol.Schema(
-            {cv.string: ha_state_detection_rules_validator(vol.Invalid)}
+# Deprecated in Home Assistant 2021.9
```
```suggestion
# Deprecated in Home Assistant 2022.2
```

ollo69 committed:
Apply suggestions from code review

Co-authored-by: Robert Hillis <tkdrob4390@yahoo.com>

bdraco commented:
Second opinion request is for
https://github.com/home-assistant/core/pull/54444/files#r767289491

bdraco commented on homeassistant/components/androidtv/config_flow.py:
```diff
@@ -0,0 +1,300 @@
+        """Attempt to connect the Android TV."""
+
+        try:
+            aftv, message = await async_connect_androidtv(self.hass, user_input)
```
```suggestion
            aftv, error_message = await async_connect_androidtv(self.hass, user_input)
```

bdraco replied:
Took a minute to see what was going on here.  Small suggestion on naming

bdraco commented on homeassistant/components/androidtv/config_flow.py:
```diff
@@ -0,0 +1,300 @@
+            return RESULT_UNKNOWN, None
+
+        if not aftv:
+            _LOGGER.warning(message)
```
```suggestion
            _LOGGER.warning(error_message)
```

bdraco commented on homeassistant/components/androidtv/config_flow.py:
```diff
@@ -0,0 +1,300 @@
+        for entry in self._async_current_entries():
+            if entry.data[CONF_HOST] == import_config[CONF_HOST]:
+                _LOGGER.warning(
+                    "Already configured. This yaml configuration has already been imported. Please remove it"
```
Would be nice to tell them for which host, but it doesn't really matter since its going to go away anyways

bdraco commented on homeassistant/components/androidtv/__init__.py:
```diff
@@ -1 +1,214 @@
+
+def _setup_androidtv(hass, config):
+    """Generate an ADB key (if needed) and load it."""
+    adbkey = config.get(CONF_ADBKEY, hass.config.path(STORAGE_DIR, "androidtv_adbkey"))
```
Is it a problem if multiple devices are using the same key?

ollo69 replied:
No, as I now. I have 2 FireTV and they use same key.

bdraco reviewed (commented):
Looks like the only material item is to get another set of eyes on https://github.com/home-assistant/core/pull/54444#discussion_r767289491

~~..unless we need a unique key per device https://github.com/home-assistant/core/pull/54444#discussion_r767293293~~

JeffLIrion commented:
> .. unless we need a unique key per device [#54444 (comment)](https://github.com/home-assistant/core/pull/54444#discussion_r767293293)

@bdraco the comment link isn't working for me, but as @ollo69 said, the same ADB key can be used for multiple devices.

ollo69 committed:
Appy naming suggestion

bdraco commented on homeassistant/components/androidtv/media_player.py:
```diff
@@ -168,180 +168,110 @@
+    ):
+        if key in config:
+            if key == CONF_STATE_DETECTION_RULES:
+                config_options[key] = json.dumps(config[key])
```
There wasn't much feedback on this topic besides there isn't a need to json encode this since its going to be stored in json anyways so we end up with double encoded json in `.storage/core.config_entries`.

ollo69 replied:
You are suggesting to save this as JSON in `config_entry` and so convert the string field in json before save it in the `options`?

bdraco replied:
I think you can drop the `json.dumps` and `json.loads` and store the data structure directly in the config entry.

ollo69 replied:
If we don't convert the JSON in string we will not be able to manage it using the options text box:
 
![image](https://user-images.githubusercontent.com/60491700/146660935-cd38e62f-d8ce-4504-a566-9c450cfc1353.png)

ollo69 replied:
I think I can improve a little bit this, creating a wizard similar to the one used for the apps list to manage rules for single apps. Please wait my next commit.

bdraco replied:
I marked it as a draft.  When you are ready for another pass, please mark it ready for review

ollo69 replied:
I created the flow to configure the detection rules, looks more simple to manage in this way.
But I found a strange behavior that I don't think is related to this PR: when I call the method `self.async_create_entry` to save changed options, if I only change `apps` or `detection_rule` (that are managed with separated flow and saved as dict), not always new values are written in `.storage/core.config_entries`. It could be a bug or I'm missing something?

bdraco replied:
Try making a `deepcopy` of the rules before modifying them. You might be updating the existing data structure which would prevent it from seeing it as new

ollo69 replied:
Of course, it was enough copy the 2 variable defined in init. Now it is ok!

ollo69 committed:
Fix wrong log message

bdraco commented:
Convert it right before you output it and when you read it back in so it’s only happening at the UI layer 

Sent from my iPhone

> On Dec 18, 2021, at 8:01 PM, ollo69 ***@***.***> wrote:
> 
> ﻿
> @ollo69 commented on this pull request.
> 
> In homeassistant/components/androidtv/media_player.py:
> 
> > +            config_data[key] = config[key]
> +
> +    # get options
> +    config_options = {}
> +    for key in (
> +        CONF_APPS,
> +        CONF_EXCLUDE_UNNAMED_APPS,
> +        CONF_GET_SOURCES,
> +        CONF_SCREENCAP,
> +        CONF_STATE_DETECTION_RULES,
> +        CONF_TURN_OFF_COMMAND,
> +        CONF_TURN_ON_COMMAND,
> +    ):
> +        if key in config:
> +            if key == CONF_STATE_DETECTION_RULES:
> +                config_options[key] = json.dumps(config[key])
> If we don't convert the JSON in string we will not be able to manage it using the options text box:
> 
> 
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

ollo69 committed:
Save detection rules as JSON in config entry

ollo69 committed:
Fix detection rules and option flow tests

ollo69 committed:
Add option flow to configure state detection rules

bdraco commented:
That ends up going through async_update_entry in homeassistant.config_entries.

If the below code doesn’t detect it as changed thats an issue:

        if options is not UNDEFINED and entry.options != options:
            changed = True
            entry.options = MappingProxyType(options)


> On Dec 20, 2021, at 10:50, ollo69 ***@***.***> wrote:
> 
> 
> @ollo69 commented on this pull request.
> 
> In homeassistant/components/androidtv/media_player.py <https://github.com/home-assistant/core/pull/54444#discussion_r772520927>:
> 
> > +            config_data[key] = config[key]
> +
> +    # get options
> +    config_options = {}
> +    for key in (
> +        CONF_APPS,
> +        CONF_EXCLUDE_UNNAMED_APPS,
> +        CONF_GET_SOURCES,
> +        CONF_SCREENCAP,
> +        CONF_STATE_DETECTION_RULES,
> +        CONF_TURN_OFF_COMMAND,
> +        CONF_TURN_ON_COMMAND,
> +    ):
> +        if key in config:
> +            if key == CONF_STATE_DETECTION_RULES:
> +                config_options[key] = json.dumps(config[key])
> I created the flow to configure the detection rules, looks more simple to manage in this way.
> But I found a strange behavior that I don't think is related to this PR: when I call the method self.async_create_entry to save changed options, if I only change apps or detection_rule (that are managed with separated flow and saved as dict), not always new values are written in .storage/core.config_entries. It could be a bug or I'm missing something?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/home-assistant/core/pull/54444#discussion_r772520927>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAFB7CD27XAV25LXRO66JMLUR5NEPANCNFSM5B5MBDKA>.
> You are receiving this because you were mentioned.
>

ollo69 committed:
Deepcopy option flow dict attributes

bdraco commented:
@ollo69 Thanks. I just realized we are missing a docs PR for this though right after I merged it. Please open one thanks 👍

frenck commented:
@ollo69 I don't see a documentation PR? Could you please create one?

ollo69 commented:
Yes, of course, I'll do on next days, just give me some times...

bdraco commented:
Thanks for opening the docs PR @ollo69  👍

MartinHjelmare commented on tests/components/androidtv/test_media_player.py:
```diff
@@ -130,14 +171,16 @@ async def _test_reconnect(hass, caplog, config):
     ], patchers.PATCH_KEYGEN, patchers.PATCH_ANDROIDTV_OPEN, patchers.PATCH_SIGNER:
-        assert await async_setup_component(hass, DOMAIN, config)
+        assert await hass.config_entries.async_setup(config_entry.entry_id)
+        # assert await async_setup_component(hass, DOMAIN, config)
```
Commented code.

MartinHjelmare commented on tests/components/androidtv/test_media_player.py:
```diff
@@ -724,185 +785,142 @@ async def test_setup_fail_firetv(hass):
-        state = hass.states.get(entity_id)
-        assert state is not None
-        assert state.attributes["adb_response"] is None
+            # patch_shell.assert_called_with(command)
```
Commented code.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!
